datafusion/functions/src/string/concat.rs

```diff
@@ -64,13 +66,19 @@ impl ScalarUDFImpl for ConcatFunc {
         &self.signature
     }
 
-    fn return_type(&self, _arg_types: &[DataType]) -> Result<DataType> {
-        Ok(Utf8)
+    fn return_type(&self, arg_types: &[DataType]) -> Result<DataType> {
+        use DataType::*;
+        Ok(match &arg_types[0] {
+            Utf8View => Utf8View,
+            LargeUtf8 => LargeUtf8,
+            _ => Utf8,
+        })
```

Comment:
the logic seems to assume all arguments are of the same type?

also, why not always return `Utf8`?
the code performing actual concatenation seems to be always the same.

Reply:
Ah yeah @findepi I think the logic is "Whatever the first argument type is the output should be of that type" so if the received values were: Utf8, Utf8View the output would be Utf8. I'm taking the logic from other UDFs and applying it here. It may not be the best way of doing this though.

Reply:
> I think the logic is "Whatever the first argument type is the output should be of that type

i understand this is what's implemented. but not sure why it is so.
what's the exact benefit of presenting the data as string view, if we computed the exact string anyway, and we technically don't need to represent it as a string view?

Reply:
yes, that's my intuition
except that i would keep building StringArray and just declare return type as Utf8 always

from the issue https://github.com/apache/datafusion/issues/11836

> Currently, a call to CONCAT with a Utf8View datatypes induces a cast. After the change that fixes this issue, it should not.

this is about _inputs_ to the function, not the return type

Side note: 
String view could be an interesting return type if we wanted to optimize for single non-null string view input and let it pass-through; but the code doesn't do this today, not sure it's worth implementing for this edge case and it should be independent of arguments order, ie not tied to the first input's type.
end of side note.

Reply:
Great that sounds good with me I'll go ahead and adjust the code.

Reply:
@findepi what about for LargeUtf8? I suspect that if a LargeUtf8 is the input then the output should also be that since its an i64 datatype vs the i32 datatype for Utf8?

Reply:
i don't know the exact rules for how we handle LargeUtf8.
i focused on the string view portion of the PR. from "string views perspective", LargeUtf8 is non-issue, so IMO it's fine not to change the return type with respect to LargeUtf8 in this PR. but i agree that we probably should return LargeUtf8 when any input is LargeUtf8 (pr what exactly the logic should be).

in fact, what does the binary concat operator do?

Reply:
@findepi when you say binary concat operator are you talking about `||` as an operator?

Reply:
> when you say binary concat operator are you talking about `||` as an operator?

yes


> the logic seems to assume all arguments are of the same type?
> 
> also, why not always return `Utf8`? the code performing actual concatenation seems to be always the same.

the question still holds (why exactly we bias towards the first param type), but i am no longer convinced about my suggestion to use Utf8 always.

i think we should "just" make sure `concat(a, b, c)`  is type-equivalent to `a || b || c`.
the `||`'s logic apparently is
- if any of the operands are Utf8View, the result is Utf8View
- else, if any of the operands are LargeUtf8, the result is LargeUtf8

cc @alamb